### DISCUSSION TITLE
Representing Bipartite graphs with concatenated features

### DISCUSSION THREAD

**Discussion by ebrahimpichka** [pyg-team/pytorch_geometric#10264] [Q&A]
Hello PyG team/community,

I have issues with working with bipartite graphs that i'd appreciate your help with. 

Suppose we have two separate sets of nodes which form our **Undirected Bipartite** graph i.e. $V =  ( V_{src}, V_{dst} ) $.

The [docs](https://pytorch-geometric.readthedocs.io/en/2.5.2/tutorial/create_gnn.html) mention that 

> For bipartite graphs with two independent sets of nodes and indices, and each set holding its own information, this split can be marked by passing the information as a tuple, e.g. x=(x_N, x_M)

## First question
My first question is, would it make a difference if we just construct one graph with (x_N, x_M) stacked together and make sure that node indices of `dst` starts after that of `src`. Like:

```py
num_src_nodes = len(x_src)
num_dst_nodes = len(x_dst)

src_indices = torch.tensor([src for src in range(num_src_nodes) for _ in range(num_dst_nodes)])
dst_indices = torch.tensor([dst for _ in range(num_src_nodes) for dst in range(num_dst_nodes)])

# Shift dst indices to ensure they're distinct from src
# edge_index with shape [2, num_edges]
edge_index = torch.stack([src_indices, dst_indices + num_src_nodes])
    
# single feature mat
x = torch.cat([x_src, x_dst], dim=0)
    
data = Data(
    x=x,
    edge_index=edge_index,
    )
```

![image](https://github.com/user-attachments/assets/b7cfc460-4010-40ef-ad64-5969d122e413)

## Second question
If there is indeed a difference, what's the advantage of separating `src` and `dst` for conv layers?
Also, is there an example of how using the `x=(x_N, x_M)` method works? i mean the entire steps of preparing the data, defining the network, propagation, etc.
